### DISCUSSION TITLE
Inquiring about Consistency of ASP.NET Core Across Supported Operating Systems

### DISCUSSION THREAD

**Discussion by phadkeSharanMatrixComsec** [dotnet/aspnetcore#55682] [Q&A]
Hi everyone,

I'm reaching out to gather insights and experiences regarding the consistency of ASP.NET Core across the various operating systems it supports. As a software developer primarily working with ASP.NET Core, I'm interested in understanding any potential challenges or issues that developers might have encountered when deploying and running ASP.NET Core applications across different operating systems.

While ASP.NET Core is known for its cross-platform capabilities, I recognize that there may still be nuances or discrepancies between different operating systems that could impact the development and deployment process. These could range from file system differences to networking behaviors and third-party library compatibility.

Could anyone share their experiences or observations regarding:

1. Any specific issues or challenges encountered when developing or deploying ASP.NET Core applications on different operating systems?
2. Strategies or best practices for ensuring consistency and compatibility across platforms?
3. Tools or resources that have been helpful in addressing platform-specific issues?
4. Inconsistencies or challenges encountered when integrating third-party APIs with ASP.NET Core applications on different operating systems?

Looking forward to hearing your insights and tips!

**Comment by davidfowl** [comment]
This isn't an exhaustive list but maybe it'll give you a high-level sense of the differences. Most of the differences are at the operation system level itself..NET does a really good job of abstracting that away from the user, but it will bleed through in various ways:

> Any specific issues or challenges encountered when developing

- Paths are case sensitive on Linux and Mac.
- You have to use C# Dev Kit on VS Code or JetBrains Rider when developing outside of windows.
- The dotnet dev certs tool for ASP.NET Core doesn't work on Linux distros (use https://github.com/tmds/linux-dev-certs)

> deploying ASP.NET Core applications on different operating systems?
 
- Paths are case sensitive on Linux. 
- There's no IIS (nor web deploy). This is one of the most significant ones. If you want to host "multiple sites" on the same compute then you'll need to use a reverse proxy like YARP (https://microsoft.github.io/reverse-proxy/) to route traffic on port 80 to other kestrel instances running on the compute. There's also nothing that will reboot your applications if they crash. Many people use systemd (https://swimburger.net/blog/dotnet/how-to-run-a-dotnet-core-console-app-as-a-service-using-systemd-on-linux) as their process launcher.
- `:` is supported in environment variables on Linux. This is why ASP.NET Core configuration uses `__` for environment variables (https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0#non-prefixed-environment-variables)
- Windows auth won't *just work* (https://learn.microsoft.com/en-us/aspnet/core/security/authentication/windowsauth#kestrel)
- The TLS (HTTPS) implementation is completely different on Linux distros, it uses open SSL. There's no standard certificate store on most linux distros so configuring certificates can vary depending on which OS you pick. You're most likely managing files.

> Strategies or best practices for ensuring consistency and compatibility across platforms?

One of the best ways to do this is to run in a linux VM. If you're a windows user then you can use WSL (https://learn.microsoft.com/en-us/windows/wsl/setup/environment). There's support in visual studio for launching your application under WSL (https://learn.microsoft.com/en-us/visualstudio/debugger/debug-dotnet-core-in-wsl-2?view=vs-2022).

> Inconsistencies or challenges encountered when integrating third-party APIs with ASP.NET Core applications on different operating systems?

Any that are tied to Windows. Anything that does native interop will be a problem unless it's multiplatform. It's hard to know this off the bat but maybe tools like https://www.ndepend.com/ can help you understand what your dependencies are doing.

> Tools or resources that have been helpful in addressing platform-specific issues?

.NET comes with https://learn.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer. We flagged sets of API that do no work cross platform so that you are aware at compile time. This isn't full proof though as its about APIs that you are using, not your dependencies.

**Reply by phadkeSharanMatrixComsec** [reply]
Thank you so much for your insight!
